core/db/migrate/20150819154308_add_discontinued_to_products_and_variants.rb

```diff
@@ -0,0 +1,66 @@
+class AddDiscontinuedToProductsAndVariants < ActiveRecord::Migration
+  def up
+    add_column :spree_products, :discontinue_on, :datetime, after: :available_on
+    add_column :spree_variants, :discontinue_on, :datetime, after: :deleted_at
+
+    puts "Warning: This migration changes the meaning of 'deleted'. Before this change, 'deleted' meant products that were no longer being sold in your store. After this change, you can only delete a product or variant if it has not already been sold to a customer (a model-level check enforces this). Instead, you should use the new field 'discontinue_on' for products or variants which were sold in the past but no longer for sale. This fixes bugs when other objects are attached to deleted products and variants. (Even though acts_as_paranoid gem keeps the records in the database, most associations are automatically scoped to exclude the deleted records.) In thew meaning of 'deleted,' you can still use the delete function on products & variants which are *truly user-error mistakes*, specifically before an order has been placed or the items have gone on sale. You also must use the soft-delete function (which still works after this change) to clean up slug (product) and SKU (variant) duplicates. Otherwise, you should generally over ever need to discontinue products.
+
+Data Fix: We will attempt to reverse engineer the old meaning of 'deleted' (no longer for sale) to the new database field 'discontinue_on'. However, since Slugs and SKUs cannot be duplicated on Products and Variants, we cannot gaurantee this to be foolproof if you have deteled Products and Variants that have duplicate Slugs or SKUs in non-deleted records. In these cases, we recommend you use the additional rake task to clean up your old records (see rake db:fix_orphan_line_items). If you have such records, this migration will leave them in place, preferring the non-deleted records over the deleted ones. However, since old line items will still be associated with deleted objects, you will still the bugs in your app until you run:
+
+rake db:fix_orphan_line_items
+
+We will print out a report of the data we are fixing now: "
+
+
+    Spree::Product.only_deleted.each do |product|
+      # determine if there is a slug duplicate
+      the_dup =  Spree::Product.find_by(slug: product.slug)
+      if the_dup.nil?
+        # check to see if there are line items attached to any variants
+        if Spree::Variant.with_deleted.where(product_id: product.id).collect{|v| v.line_items}.any?
```

Comment:
Prefer `map` over `collect`.<br>Space missing to the left of {.<br>Space between { and | missing.<br>Space missing inside }.<br>Pass `&:line_items` as an argument to `collect` instead of a block.